### PR TITLE
py-numpy: build in parallel

### PR DESCRIPTION
**Pull request by adamjstewart**
@rgommers can you see if this improves build times? May have to explicitly convert `make_jobs` to a string, I can never remember. This can be controlled from the command line using `spack install -j#`, so `-j1` for serial or `-j16` for parallel. By default, Spack uses the number of cores on the system (capped at 16 I think).

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @adamjstewart. This didn't help unfortunately. I misremembered, and the numpy docs are wrong on this. There's two types of parallelism:
1. The `-j` (`--parallel`) option to `python setup.py install`. This is the number of extension modules to build in parallel.
2. The number of threads to use for a single Python extension. This is controlled by either `-j` or by `NPY_NUM_BUILD_JOBS`. If not set, it already defaults to `min(cpu_count, 8)`.

(1) helps more than (2). Setting `NPY_NUM_BUILD_JOBS` to `8` builds in 2 minutes. Using `-j8` builds in 1 minute.

There is no env var to control (1). If there is no way to change the build command to `python setup.py install -j $(make_jobs)`, it's best to leave it as is.

**Comment by adamjstewart** [comment]
Thanks @rgommers. We can easily add `-j#` to the `python setup.py install` command. Unfortunately, there's no way to do this automatically for all packages that depend on numpy, it would need to be added on a package-by-package basis.

Before I add this, I will say I've had some issues with this flag in the past. First of all, many packages like scipy support `python setup.py build -j8` or `python setup.py build_ext -j8` but not `python setup.py install -j8` due to the way that flags are forwarded. Spack no longer has a separate `python setup.py build` phase so we can only add this flag if it is supported by the install step. Also, I've found that this flag doesn't work even during build-time for packages like scipy: https://github.com/scipy/scipy/issues/7112. Has anything changed about this?

Also, if the above concerns are no longer issues, a few additional questions:

1. Which versions of numpy/scipy/etc actually support this flag at install time?
2. Should we add both `-j#` and `NPY_NUM_BUILD_JOBS=#` for maximum parallelism, or set `-j#` and `NPY_NUM_BUILD_JOBS=1` to prevent parallelism beyond the number of cores?
3. Which other packages should we investigate adding this flag to? I feel like I've tried adding this flag during build-time to numpy/scipy/pandas in the past, not sure what other packages might benefit.

**Comment by adamjstewart** [comment]
@rgommers any comments on the above?

**Comment by rgommers** [comment]
> Spack no longer has a separate `python setup.py build` phase so we can only add this flag if it is supported by the install step.

Okay, I tried that and unfortunately it's a showstopper, it ends with `error: option -j not recognized`. I didn't know that because I never use `setup.py install` anymore. 

> Also, I've found that this flag doesn't work even during build-time for packages like scipy: [scipy/scipy#7112](https://github.com/scipy/scipy/issues/7112). Has anything changed about this?

It's indeed still disabled for SciPy; the Cython codegen build will run in parallel still, but that won't save _that_ much time. For NumPy it's not disabled, nor for any other packages I'm aware. The race condition is in the Fortran support, which not many packages other than SciPy have.

SciPy 1.9.0 will default to building with Meson, and that will build in parallel. By default Ninja will use `2*n_cores + 2` threads.

tl;dr I think it's best to close this. The `install -j` option would have helped for NumPy the most, but it doesn't work. And once we port NumPy to Meson, the problem will be solved. Sorry about the noise.

**Comment by adamjstewart** [comment]
I'm not a huge fan of requiring non-Python deps like meson to build but hopefully it won't be too difficult to integrate properly.

**Comment by rgommers** [comment]
It should be easy - it's not a non-Python dep: https://pypi.org/project/meson/. Also written in pure Python (okay, does need Ninja and pkg-config, but it's a major upgrade from `distutils`).

**Comment by eli-schwartz** [comment]
Not entirely sure I understand the generic issue with non-python deps, since GCC/clang is also not written in python. But if it's very important to use python deps, we're indeed halfway there since Meson is written in python. The next step would be for someone to write a ninja reimplementation in python.

**Comment by adamjstewart** [comment]
> Not entirely sure I understand the generic issue with non-python deps, since GCC/clang is also not written in python.

Compilers usually come with the system, or are installed with binaries instead of from source, so they aren't that big of a deal. Meson/ninja aren't that bad either since they are relatively quick to build. My bigger issue comes from a lot of packages adopting dependencies on rust (setuptools-rust) and npm (node-js) since these can take hours to build from source. 

I've also had a lot of problems with libraries like rasterio/fiona which depend on GDAL/PROJ but don't build them, so a simple `pip install rasterio` won't work in CI unless these libs are already installed or unless wheels are available. For these particular examples, wheels aren't available for Python 3.10 or for Windows, so I have to use conda. Hopefully someday Spack will have binaries for all these packages

**Comment by eli-schwartz** [comment]
> My bigger issue comes from a lot of packages adopting dependencies on rust (setuptools-rust) and npm (node-js) since these can take hours to build from source.

Ah, so less "non-python" and more "giant honking additional languages that take a surprisingly long time to build anything, especially to build the compiler".

Yeah, that can be a bit frustrating indeed. The problems don't even end once you have a rust binary, a setuptools-rust package still takes ages to build after all dependencies are built and installed. And on top of that, cargo basically cannot be parallelized by design. (Meson has draft/WIP support for `rustc` without `cargo` if you really want it, which will have actual real parallelism.)

**Comment by adamjstewart** [comment]
Yeah, cargo also likes to download and install its own deps, which makes it really difficult to control on an air-gapped system without internet access. We're still working on better rust/cargo support: #16185
